File: lintcode/python/0206_interval_sum.py
#!/usr/bin/python -t

# segment tree and binary search
# build O(n) query O(logn), space O(n)

"""
Definition of Interval.
class Interval(object):
    def __init__(self, start, end):
        self.start = start
        self.end = end
"""

class sgtree(object):
    def __init__(self, start, end, s_val=0):
        self.start = start
        self.end = end
        self.s_val = s_val
        self.left, self.right = None, None
        
    @classmethod
    def build(cls, start, end, s):
        if start > end:
            return None
        if start == end:
            return sgtree(start, end, s[start])
        
        node = sgtree(start, end, s[start])
        mid = (start+end)/2
        node.left = cls.build(start, mid, s)
        node.right = cls.build(mid+1, end, s)
        node.s_val = node.left.s_val + node.right.s_val
        
        return node
        
    @classmethod
    def query(cls, root, start, end):
        if root.start > end or root.end < start:
            return 0
            
        if start <= root.start and root.end <= end:
            return root.s_val
            
        return (cls.query(root.left, start, end)+cls.query(root.right, start, end))
        

class Solution:
    """
    @param A: An integer list
    @param queries: An query list
    @return: The result list
    """
    def intervalSum(self, A, queries):
        # write your code here
        root = sgtree.build(0, len(A)-1, A)
        ret = []
        for q in queries:
            ret.append(sgtree.query(root, q.start, q.end))
            
        return ret

File: lintcode/golang/0206_interval_sum.go
/**
 * Definition of Interval:
 * type Interval struct {
 *     Start, End int
 * }
 */

/**
 * @param A: An integer list
 * @param queries: An query list
 * @return: The result list
 */
 
type SegNode struct {
    Start, End int 
    Total int64
    Left, Right *SegNode
}

func buildSeg (a []int, start, end int) *SegNode {
    if start > end {
        return nil
    }
    if start == end {
        return &SegNode {
            Start: start,
            End: end,
            Total: int64(a[start]),
        }
    }
    
    root := SegNode {
        Start: start,
        End: end,
    }
    
    mid := (start+end)/2
    root.Left = buildSeg(a, start, mid)
    root.Right = buildSeg(a, mid+1, end)
    root.Total = root.Left.Total + root.Right.Total
    
    return &root
}

func querySeg (node *SegNode, start, end int) int64 {
    if start > node.End || end < node.Start {
        return 0
    }
    if start <= node.Start && end >= node.End {
        return node.Total
    }
    
    return querySeg(node.Left, start, end) + querySeg(node.Right, start, end)
}

func intervalSum (A []int, queries []*Interval) []int64 {
    // write your code here
    var ret []int64
    
    root := buildSeg(A, 0, len(A)-1)
    
    for _, i := range queries {
        ret = append(ret, querySeg(root, i.Start, i.End))
    }
    
    return ret
}

